File: Sources/WInstallerApp/AssistantModel.swift
import Foundation
import SwiftUI
import WInstallerCore

/// Coordinates the assistant flow: owns user intent, confirmation state, live
/// device data, and the running operation. Long work runs in cancellable tasks;
/// UI-facing state stays on the main actor.
@MainActor
final class AssistantModel: ObservableObject {
    enum Step: Int, CaseIterable, Identifiable {
        case welcome = 1
        case chooseISO
        case verifyISO
        case insertUSB
        case analyzeUSB
        case confirmErase
        case createUSB
        case done
        case vmware

        var id: Int { rawValue }

        var title: String {
            switch self {
            case .welcome: "Welcome"
            case .chooseISO: "Choose ISO"
            case .verifyISO: "Verify ISO"
            case .insertUSB: "Insert USB"
            case .analyzeUSB: "Analyze USB"
            case .confirmErase: "Confirm Erase"
            case .createUSB: "Create USB"
            case .done: "Done"
            case .vmware: "VMware Fusion"
            }
        }

        var symbol: String {
            switch self {
            case .welcome: "sparkles"
            case .chooseISO: "opticaldisc"
            case .verifyISO: "checkmark.seal"
            case .insertUSB: "cable.connector"
            case .analyzeUSB: "list.bullet.clipboard"
            case .confirmErase: "exclamationmark.triangle"
            case .createUSB: "gearshape.2"
            case .done: "checkmark.circle"
            case .vmware: "macwindow.on.rectangle"
            }
        }
    }

    // Flow state
    @Published var step: Step = .welcome
    @Published var iso: InstallerISO?
    @Published var selectedDrive: USBDrive?
    @Published var plan: OperationPlan?
    @Published var events: [EngineEvent] = []
    @Published var confirmationText = ""
    @Published var errorMessage: String?

    // UI state
    @Published var showingFileImporter = false
    @Published var showingTechnicalDetails = false
    @Published var isInspectingISO = false
    @Published var isRefreshingDrives = false
    @Published var isRunning = false
    @Published var simulateMode = false

    // Device data
    @Published var drives: [USBDrive] = []
    @Published var vmwareApps: [VirtualizationApp] = []
    @Published var commandLog = ""

    private let engine = BootableUSBEngine()
    private let liveRunner: CommandRunning = ProcessCommandRunner()
    private let logger = LocalLogger()
    private let detector = VirtualizationDetector()
    private var executionTask: Task<Void, Never>?

    var canContinue: Bool {
        switch step {
        case .welcome: true
        case .chooseISO: iso != nil && !isInspectingISO
        case .verifyISO: (iso?.confidence ?? .low) >= .medium
        case .insertUSB: selectedDrive != nil
        case .analyzeUSB: plan != nil
        case .confirmErase: confirmationText == selectedDrive?.displayName
        case .createUSB: !isRunning
        case .done, .vmware: true
        }
    }

    // MARK: - Navigation

    func continueFlow() {
        errorMessage = nil
        switch step {
        case .welcome:
            step = .chooseISO
        case .chooseISO:
            step = .verifyISO
        case .verifyISO:
            step = .insertUSB
            refreshDrives()
        case .insertUSB:
            createPlan()
            if plan != nil { step = .analyzeUSB }
        case .analyzeUSB:
            step = .confirmErase
        case .confirmErase:
            confirmAndCreate()
        case .createUSB:
            step = .done
        case .done:
            detectVMware()
            step = .vmware
        case .vmware:
            reset()
        }
    }

    func goBack() {
        guard step.rawValue > Step.welcome.rawValue, step.rawValue < Step.createUSB.rawValue else { return }
        step = Step(rawValue: step.rawValue - 1) ?? .welcome
    }

    // MARK: - ISO

    func importISO(result: Result<[URL], Error>) {
        switch result {
        case .success(let urls):
            guard let url = urls.first else { return }
            guard url.pathExtension.lowercased() == "iso" else {
                errorMessage = "Choose a file with the .iso extension."
                return
            }
            inspectISO(url)
        case .failure(let error):
            errorMessage = error.localizedDescription
        }
    }

    private func inspectISO(_ url: URL) {
        isInspectingISO = true
        errorMessage = nil
        Task {
            let access = url.startAccessingSecurityScopedResource()
            defer { if access { url.stopAccessingSecurityScopedResource() } }
            let inspector = ISOInspector(runner: liveRunner, engine: engine)
            do {
                iso = try await inspector.inspect(url: url)
            } catch let error as ISOInspectionError {
                errorMessage = error.userMessage
            } catch let error as BootableUSBError {
                errorMessage = error.userMessage
            } catch {
                errorMessage = error.localizedDescription
            }
            isInspectingISO = false
        }
    }

    // MARK: - USB

    func refreshDrives() {
        isRefreshingDrives = true
        Task {
            let enumerator = DiskEnumerator(runner: liveRunner)
            let found = (try? await enumerator.removableDrives()) ?? []
            drives = found
            if let selectedDrive, !found.contains(where: { $0.id == selectedDrive.id }) {
                self.selectedDrive = nil
            }
            isRefreshingDrives = false
        }
    }

    private func createPlan() {
        guard let iso, let selectedDrive else { return }
        do {
            let created = try engine.makePlan(iso: iso, drive: selectedDrive, tools: ToolAvailability(wimlibImageX: true))
            plan = created
            events = seedEvents(for: created)
        } catch let error as BootableUSBError {
            errorMessage = error.userMessage
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Execution

    private func confirmAndCreate() {
        guard let plan, let selectedDrive else { return }
        do {
            try engine.confirmErase(for: selectedDrive, typedName: confirmationText)
        } catch let error as BootableUSBError {
            errorMessage = error.userMessage
            return
        } catch {
            errorMessage = error.localizedDescription
            return
        }

        step = .createUSB
        isRunning = true
        errorMessage = nil
        events = seedEvents(for: plan)

        let executor = simulateMode
            ? Simulation.executor(for: plan, logger: logger)
            : OperationExecutor(runner: liveRunner, logger: logger)

        executionTask = Task {
            do {
                for try await event in executor.run(plan: plan) {
                    apply(event)
                }
                isRunning = false
                commandLog = await logger.exportText()
                step = .done
            } catch let error as ExecutionError {
                isRunning = false
                errorMessage = error.userMessage
                commandLog = await logger.exportText()
            } catch {
                isRunning = false
                errorMessage = error.localizedDescription
            }
        }
    }

    func cancel() {
        executionTask?.cancel()
        executionTask = nil
        isRunning = false
    }

    private func apply(_ event: EngineEvent) {
        if let index = events.firstIndex(where: { $0.id == event.id }) {
            events[index] = event
        } else {
            events.append(event)
        }
    }

    private func seedEvents(for plan: OperationPlan) -> [EngineEvent] {
        var seeds: [(String, String)] = [
            ("mount-iso", "Mount ISO read-only"),
            ("verify-usb", "Re-check USB identity"),
            ("unmount", "Unmount USB drive"),
            ("erase-disk", "Erase and format USB drive"),
            ("copy-files", "Copy installer files")
        ]
        if plan.strategy.requiresWIMSplit {
            seeds.append(("split-wim", "Split oversized Windows image"))
        }
        seeds.append(contentsOf: [
            ("validate", "Validate boot files"),
            ("eject", "Eject USB safely")
        ])
        return seeds.map { EngineEvent(id: $0.0, state: .idle, title: $0.1, detail: "Waiting…", status: .waiting) }
    }

    // MARK: - VMware

    func detectVMware() {
        vmwareApps = detector.detectAll()
    }

    var vmwareFusion: VirtualizationApp? {
        vmwareApps.first { $0.id == VirtualizationTarget.vmwareFusion.bundleIdentifier }
    }

    func openVMware() {
        guard let app = vmwareFusion, app.isInstalled else { return }
        detector.open(app)
    }

    // MARK: - Reset

    private func reset() {
        cancel()
        engine.reset()
        step = .welcome
        iso = nil
        selectedDrive = nil
        plan = nil
        events = []
        confirmationText = ""
        errorMessage = nil
        commandLog = ""
    }

    #if DEBUG
    /// Loads a fabricated Windows ISO so the UI can be exercised without hardware.
    func loadSampleISO() {
        do {
            iso = try engine.analyzeISO(
                url: URL(filePath: "/Users/example/Downloads/Win11_English_x64.iso"),
                size: 6_500_000_000,
                volumeLabel: "CCCOMA_X64FRE",
                directoryEntries: ["setup.exe", "sources/boot.wim", "sources/install.wim", "efi/boot/bootx64.efi"],
                fileSizes: ["sources/install.wim": 5_100_000_000]
            )
        } catch let error as BootableUSBError {
            errorMessage = error.userMessage
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func loadSampleDrives() {
        drives = [.sampleRemovable]
    }
    #endif
}

File: apps/windows/WInstaller.App/AssistantModel.cs
using WInstaller.Core;

namespace WInstaller.App;

public enum AssistantStep
{
    Welcome = 1,
    ChooseIso,
    VerifyIso,
    InsertUsb,
    AnalyzeUsb,
    ConfirmErase,
    CreateUsb,
    Done,
}

/// <summary>
/// Coordinates the assistant flow: owns user intent, confirmation state, live
/// device data, and the running operation. Mirrors the macOS AssistantModel;
/// the flow ends at Done because the virtualization-handoff step is
/// not-applicable on Windows (docs/screen-inventory.yaml).
/// </summary>
public sealed class AssistantModel
{
    /// <summary>Raised on every state transition the UI must re-render for.</summary>
    public event Action? Changed;

    // Flow state
    public AssistantStep Step { get; private set; } = AssistantStep.Welcome;
    public InstallerIso? Iso { get; private set; }
    public UsbDrive? SelectedDrive { get; private set; }
    public OperationPlan? Plan { get; private set; }
    public List<EngineEvent> Events { get; } = [];
    public string ConfirmationText { get; private set; } = "";
    public string? ErrorMessage { get; private set; }

    // UI state
    public bool IsInspectingIso { get; private set; }
    public bool IsRefreshingDrives { get; private set; }
    public bool IsRunning { get; private set; }
    public bool SimulateMode { get; private set; }

    // Device data
    public IReadOnlyList<UsbDrive> Drives { get; private set; } = [];
    public string CommandLog { get; private set; } = "";

    private readonly BootableUsbEngine _engine = new();
    private readonly ICommandRunner _liveRunner = new ProcessCommandRunner();
    private readonly LocalLogger _logger = new();
    private CancellationTokenSource? _executionCancellation;

    public bool CanContinue => Step switch
    {
        AssistantStep.Welcome => true,
        AssistantStep.ChooseIso => Iso is not null && !IsInspectingIso,
        AssistantStep.VerifyIso => Iso?.Confidence >= DetectionConfidence.Medium,
        AssistantStep.InsertUsb => SelectedDrive is not null,
        AssistantStep.AnalyzeUsb => Plan is not null,
        AssistantStep.ConfirmErase => ConfirmationText == SelectedDrive?.DisplayName,
        AssistantStep.CreateUsb => !IsRunning,
        AssistantStep.Done => true,
        _ => false,
    };

    public bool CanGoBack => Step > AssistantStep.Welcome && Step < AssistantStep.CreateUsb;

    // MARK: Navigation

    public void ContinueFlow()
    {
        ErrorMessage = null;
        switch (Step)
        {
            case AssistantStep.Welcome:
                Step = AssistantStep.ChooseIso;
                break;
            case AssistantStep.ChooseIso:
                Step = AssistantStep.VerifyIso;
                break;
            case AssistantStep.VerifyIso:
                Step = AssistantStep.InsertUsb;
                _ = RefreshDrivesAsync();
                break;
            case AssistantStep.InsertUsb:
                CreatePlan();
                if (Plan is not null)
                {
                    Step = AssistantStep.AnalyzeUsb;
                }
                break;
            case AssistantStep.AnalyzeUsb:
                Step = AssistantStep.ConfirmErase;
                break;
            case AssistantStep.ConfirmErase:
                _ = ConfirmAndCreateAsync();
                break;
            case AssistantStep.CreateUsb:
                Step = AssistantStep.Done;
                break;
            case AssistantStep.Done:
                Reset();
                break;
        }
        RaiseChanged();
    }

    public void GoBack()
    {
        if (!CanGoBack)
        {
            return;
        }
        ErrorMessage = null;
        Step = (AssistantStep)((int)Step - 1);
        RaiseChanged();
    }

    // MARK: ISO

    public async Task ImportIsoAsync(string path)
    {
        if (!path.EndsWith(".iso", StringComparison.OrdinalIgnoreCase))
        {
            ErrorMessage = "Choose a file with the .iso extension.";
            RaiseChanged();
            return;
        }

        IsInspectingIso = true;
        ErrorMessage = null;
        RaiseChanged();
        try
        {
            var inspector = new IsoInspector(_liveRunner, _engine);
            Iso = await inspector.InspectAsync(path);
        }
        catch (IsoInspectionException exception)
        {
            ErrorMessage = exception.UserMessage;
        }
        catch (BootableUsbException exception)
        {
            ErrorMessage = exception.UserMessage;
        }
        catch (Exception exception)
        {
            ErrorMessage = exception.Message;
        }
        IsInspectingIso = false;
        RaiseChanged();
    }

    // MARK: USB

    public async Task RefreshDrivesAsync()
    {
        IsRefreshingDrives = true;
        RaiseChanged();
        try
        {
            var enumerator = new DiskEnumerator(_liveRunner);
            Drives = await enumerator.RemovableDrivesAsync();
        }
        catch (Exception)
        {
            Drives = [];
        }
        if (SelectedDrive is not null && Drives.All(drive => drive.DiskNumber != SelectedDrive.DiskNumber))
        {
            SelectedDrive = null;
        }
        IsRefreshingDrives = false;
        RaiseChanged();
    }

    public void SelectDrive(UsbDrive? drive)
    {
        SelectedDrive = drive;
        RaiseChanged();
    }

    private void CreatePlan()
    {
        if (Iso is null || SelectedDrive is null)
        {
            return;
        }
        try
        {
            Plan = _engine.MakePlan(Iso, SelectedDrive);
            SeedEvents(Plan);
        }
        catch (BootableUsbException exception)
        {
            ErrorMessage = exception.UserMessage;
        }
    }

    // MARK: Confirmation input (does not raise Changed: the text box owns focus)

    public void SetConfirmationText(string text) => ConfirmationText = text;

    public void SetSimulateMode(bool enabled) => SimulateMode = enabled;

    // MARK: Execution

    private async Task ConfirmAndCreateAsync()
    {
        if (Plan is null || SelectedDrive is null)
        {
            return;
        }
        try
        {
            _engine.ConfirmErase(SelectedDrive, ConfirmationText);
        }
        catch (BootableUsbException exception)
        {
            ErrorMessage = exception.UserMessage;
            RaiseChanged();
            return;
        }

        Step = AssistantStep.CreateUsb;
        IsRunning = true;
        ErrorMessage = null;
        SeedEvents(Plan);
        RaiseChanged();

        var executor = SimulateMode
            ? Simulation.Executor(Plan, _logger)
            : new OperationExecutor(_liveRunner, logger: _logger);

        _executionCancellation = new CancellationTokenSource();
        try
        {
            await foreach (var engineEvent in executor.Run(Plan, _executionCancellation.Token))
            {
                Apply(engineEvent);
            }
            IsRunning = false;
            CommandLog = _logger.ExportText();
            Step = AssistantStep.Done;
        }
        catch (ExecutionException exception)
        {
            IsRunning = false;
            ErrorMessage = exception.UserMessage;
            CommandLog = _logger.ExportText();
        }
        catch (Exception exception)
        {
            IsRunning = false;
            ErrorMessage = exception.Message;
        }
        RaiseChanged();
    }

    public void CancelExecution()
    {
        _executionCancellation?.Cancel();
        _executionCancellation = null;
    }

    private void Apply(EngineEvent engineEvent)
    {
        var index = Events.FindIndex(existing => existing.Id == engineEvent.Id);
        if (index >= 0)
        {
            Events[index] = engineEvent;
        }
        else
        {
            Events.Add(engineEvent);
        }
        RaiseChanged();
    }

    private void SeedEvents(OperationPlan plan)
    {
        Events.Clear();
        var seeds = new List<(string Id, string Title)>
        {
            ("mount-iso", "Mount ISO read-only"),
            ("verify-usb", "Re-check USB identity"),
            ("prepare-usb", "Erase and format USB drive"),
            ("copy-files", "Copy installer files"),
        };
        if (plan.Strategy.RequiresWimSplit)
        {
            seeds.Add(("split-wim", "Split oversized Windows image"));
        }
        seeds.Add(("validate", "Validate boot files"));
        seeds.Add(("eject", "Eject USB safely"));

        Events.AddRange(seeds.Select(seed =>
            new EngineEvent(seed.Id, EngineState.Idle, seed.Title, "Waiting…", ChecklistStatus.Waiting)));
    }

    // MARK: Reset

    private void Reset()
    {
        CancelExecution();
        _engine.Reset();
        Step = AssistantStep.Welcome;
        Iso = null;
        SelectedDrive = null;
        Plan = null;
        Events.Clear();
        ConfirmationText = "";
        ErrorMessage = null;
        CommandLog = "";
        IsRunning = false;
    }

    private void RaiseChanged() => Changed?.Invoke();
}
